app/Libraries/Payments/PaymentProcessor.php

```diff
@@ -0,0 +1,280 @@
+                break;
+            case NotificationType::REFUND:
+                $this->cancel();
+                break;
+            default:
+                throw new UnsupportedNotificationTypeException($type);
+        }
+    }
+
+    /**
+     * Processes the payment transaction.
+     *
+     * @return void
+     */
+    public function apply()
+    {
+        if (!$this->validateTransaction()) {
+            $this->throwValidationFailed(new PaymentProcessorException($this->validationErrors()));
+        }
+
+        DB::connection('mysql-store')->transaction(function () {
+            $order = $this->getOrder();
+
+            // Using a unique constraint, so we don't need to lock any rows.
+            $payment = new Payment([
+                'provider' => $this->getPaymentProvider(),
+                'transaction_id' => $this->getPaymentTransactionId(),
+                'paid_at' => $this->getPaymentDate(),
+            ]);
+
+            if (!$order->payments()->save($payment)) {
+                throw new ModelNotSavedException('failed saving model');
+            }
+
+            $order->paid($payment);
+
+            $eventName = "store.payments.completed.{$payment->provider}";
+            event($eventName, new PaymentEvent($order));
+        });
+    }
+
+    /**
+     * Cancels the payment transaction.
+     *
+     * @return void
+     */
+    public function cancel()
+    {
+        if (!$this->validateTransaction()) {
+            $this->throwValidationFailed(new PaymentProcessorException($this->validationErrors()));
+        }
+
+        DB::connection('mysql-store')->transaction(function () {
+            $order = $this->getOrder();
+            $payment = $order->payments->where('cancelled', false)->first();
+            $payment->cancel();
+
+            $order->cancel();
+
+            $eventName = "store.payments.cancelled.{$payment->provider}";
+            event($eventName, new PaymentEvent($order));
+        });
+    }
+
+    public function ensureValidSignature()
+    {
+        // TODO: post many warnings
+        if (!$this->signature->isValid()) {
+            $this->validationErrors()->add('header.signature', '.signature.not_match');
+            $this->throwValidationFailed(new InvalidSignatureException());
+        }
+    }
+
+    /**
+     * Sends a ValidationFailedEvent with the validation errors.
+     *
+     * @return void
+     */
+    protected function dispatchValidationFailed()
+    {
+        event(
+            "store.payments.validation.failed.{$this->getPaymentProvider()}",
+            new ProcessorValidationFailed($this, $this->validationErrors())
+        );
+    }
+
+    /**
+     * Fetches the Order corresponding to this payment and memoizes it.
+     *
+     * @return Order
+     */
+    protected function getOrder()
+    {
+        // ...maybe use array_key_exists and an array instead? D:
+        if (!isset($this->order) && !(isset($this->order) && $this->order === null)) {
```

Comment:
the second part won't pass because variable containing `null` will have its `isset($nullvar)` false ಠ_ಠ